flaml/spark/utils.py

```diff
@@ -0,0 +1,173 @@
+        **kwargs: parameters to store in object store.
+
+    Returns:
+        A new function with partial application of the given arguments
+        and keywords. The given arguments and keywords will be broadcasted
+        to all the executors.
+
+
+    ```python
+    import pyspark
+    import flaml
+    from sklearn.datasets import load_iris
+    def train(config, data=None):
+        if isinstance(data, pyspark.broadcast.Broadcast):
+            data = data.value
+        print(config, data)
+
+    data = load_iris()
+    with_parameters_train = flaml.spark.utils.with_parameters(train, data=data)
+    with_parameters_train(config=1)
+    train(config={"metric": "accuracy"})
+    ```
+    """
+
+    if not callable(trainable):
+        raise ValueError(
+            f"`with_parameters() only works with function trainables`. "
+            f"Got type: "
+            f"{type(trainable)}."
+        )
+
+    check_spark()
+    spark = SparkSession.builder.getOrCreate()
+
+    bc_kwargs = dict()
+    for k, v in kwargs.items():
+        bc_kwargs[k] = spark.sparkContext.broadcast(v)
+
+    return partial(trainable, **bc_kwargs)
+
+
+def customize_learner(learner_code="", file_name="mylearner"):
+    """Write customized learner/metric code contents to a file for importing.
+    It is necessary for using the customized learner in spark backend.
+    The path of the learner file will be returned.
+
+    Args:
+        learner_code: str, default="" | code contents of the customized learner.
+        file_name: str, default="mylearner" | file name of the customized learner.
+
+    Returns:
+        The path of the customized learner file.
+    ```python
+    import pyspark
+    from flaml.spark.utils import customize_learner
+    from flaml.automl.model import LGBMEstimator
+
+    learner_code = '''
+    from flaml.automl.model import LGBMEstimator
+    from flaml import tune
+
+    class MyLargeLGBM(LGBMEstimator):
+        @classmethod
+        def search_space(cls, **params):
+            return {
+                "n_estimators": {
+                    "domain": tune.lograndint(lower=4, upper=32768),
+                    "init_value": 32768,
+                    "low_cost_init_value": 4,
+                },
+                "num_leaves": {
+                    "domain": tune.lograndint(lower=4, upper=32768),
+                    "init_value": 32768,
+                    "low_cost_init_value": 4,
+                },
+            }
+    '''
+
+    customize_learner(learner_code=learner_code)
+    from flaml.spark.mylearner import MyLargeLGBM
+    assert isinstance(MyLargeLGBM(), LGBMEstimator)
+    ```
+    """
+    flaml_path = os.path.dirname(os.path.abspath(__file__))
+    learner_code = textwrap.dedent(learner_code)
+    learner_path = os.path.join(flaml_path, file_name + ".py")
+
+    def writefile(_):
+        with open(learner_path, "w") as f:
+            f.write(learner_code)
+
+    # write learner into a file
+    writefile(None)
```

Comment:
Would be neater to inline this. Don't feel the nested function is adding anything and odd to have an unused argument